### ISSUE TITLE
Wire System Sharing for multiple images

### ISSUE THREAD

**Issue by MadEgg** [wireapp/wire-ios#664]
iOS allows you to select multiple images or videos to share simultanously. However, Wire doesn't respond well to this. Sharing one image works fine. When you share multiple, no visual feedback appears after a long pause and you return to the media selection screen. Meanwhile, when you open Wire it first looks like nothing is happening. After a short while (30 seconds or so) all of a sudden the images appear. However, it seems that Wire accepts a maximum of 5 images at the same time, but doesn't give any feedback when you send more, but just picks 5. It should give a popup stating that max 5 images can be shared at a time.

And the worst part: these five images do not get sent. The get a status of 'Sending' but they never are. And the state is hidden by default so you have to tap the images to find out they aren't sent. When you send new messages or images afterwards, these are delivered but the older images remain unsent.

You can get these stalled images to send by shutting down Wire and restarting it, that will trigger a new attempt apparently.

**Comment by jenskasemets** [comment]
Hi @MadEgg ,

Thank you for your report. This issue you described will be fixed in next upcoming release. I'll let you know once available.

**Comment by jenskasemets** [comment]
Hi @MadEgg, this issue should be now fixed in release 2.30, please update and test if it behaves better.

**Comment by MadEgg** [comment]
I already tested it out yesterday evening. Sending multiple images is working as it should now!

I also noticed that when I select more than 9 images, Wire is removed from the list of apps I can use to share it, and it returns when unselecting the 10th image. Very nice!

**Comment by MadEgg** [comment]
@jenskasemets Sorry, I responded a tiny bit too quickly. Sending multiple images seems to work fine. I just tested with more images, and for some reason, I could now actually select 16 images and send them at the same time. I spammed your trusty old Otto the Bot with those. My phone delivered all 16 images and shows the responses just fine. However, I also have a Linux desktop Wire client connected to my account. On this account, I do not see 16 images I sent, I see 18. Two of them are duplicates. I'm not sure if this is a bug in the iOS client or the desktop client, or the server, but it's not working 100% perfect apparently. Much better than before though!

**Comment by jenskasemets** [comment]
@MadEgg Interesting found. What happens if you switch between conversations on desktop client? Or if you restart desktop client, does it still show some images twice?

**Comment by MadEgg** [comment]
Yes it does. I tried switching conversations a couple of times, still 18 images, 2 double. After quitting and restarting, still 18 images, 2 double.

**Comment by jenskasemets** [comment]
Thank you, we'll investigate it.

**Comment by anatoli26** [comment]
May be related about duplicate images: https://github.com/wireapp/wire-desktop/issues/472

**Comment by MadEgg** [comment]
After a couple of days I checked the image conversation with Otto the Bot again, and now I also looked at the number of responses I got from Otto the Bot.

iPhone:
* 16 outgoing images
* 12 received images from Otto

Linux Desktop Client:
* 18 outgoing images
* 11 received images from Otto

I'm not sure if even with this amount of messages, Otto is supposed to reply with an image to every single one, but if it is, it seems to indicate that Otto the Bot also did not receive every message. It's particularly odd that my phone received 12 images back from Otto and my desktop only 11.

If there's any addtional information I can provide, please let me know.

**Comment by anatoli26** [comment]
Somehow related: #792

**Comment by jenskasemets** [comment]
Hi @MadEgg, please can you retest it with latest version (2.32), there were some improvements for this behaviour in release 2.31

**Comment by MadEgg** [comment]
I just tested it after updating to Wire v2.32 (2795). I created a new test-set for this purpose: 20 pictures, showing sequential numbers of 1-20.

I selected them all in order and sent them on my iPhone, accompanied with a text.

**Observations iPhone**
The text was sent first, and Otto replied to it directly.

My text is first, and then the numbers 1 - 20 in the correct order, without duplicates.
Otto's response: 1 image after the text, 17 images following my images.

Full conversation transcription (T is my text message, 1-20 are my images, O0 is Otto's response to the text message, O1-O17 are Otto's responses afterwards).

T, O0, 1, 2, O1, 3, O2, 4, O3, 5, 6, O4, 7, 8, O5, 9, 10, O6, O7, 11, O8, 12, 13, O9, 14, O10, O11, 15, O12, 16, O13, 17, 18, O14, O15, 19, 20, O16, O17

So, in total 21 message sent, 18 responses received.


**Observations on Wire Desktop Linux:**
My text is first, and then Otto's response.
Then:

T, O0, 1, 1*, 2, O1, 3, 3*, O2, 4, O3, 5, 6, O4, 7, 8, O5, 9, O6, 10, O7, 11, O8, 12, O9, 13, O10, 14, 14*, O11, 15, O12, 16, O13, 17, 18, O14, 18*, O15, 19, O16, 20, O17.

The * indicate duplicate images. So in total, it looks like I sent 25 messages, including 4 duplicate images. Otto responded to the text message and 17 images.

**Summarizing**
It seems Otto still did not receive all my messages, or at least did not respond to all of them. My desktop client receives all images in the correct order, but adds some duplicates here and there.

Still not 100% correct. In real-world usage, sending real images to my contacts, I've encountered no such problems, by the way. But then again: I never spam them with 20 images at the same time ;-)

**Comment by jenskasemets** [comment]
@MadEgg there were some improvements to share extension behaviour on just released version 2.33. Would you be so kind and re-test your previous scenario and let us know if you still see any issues or not.

**Comment by MadEgg** [comment]
I tried it again. Unfortunately, this attempt was worse. First of all, the sending screen would not go away. I'm talking about the screen here that pops up when invoking the System Sharing system from the photo gallery, after I select the images and Wire, and the recipient. However, it did transmit several images. After 5 minutes of not having received additional images on my desktop, I terminated the photo gallery as it did not seem that anything was actually still going on.

Results:

On my phone, I see:

T, 1, 2, O0, 3, 4, O1, X, X, O2, X, O3, 8, 9, 10, O4, 11, O5, 12, 13, O6, 14, O7, 15, 16, O8, X, 18, 19, 20, O9, O10, O11.

X here means, image not loading. It just shows the rotating loading icon, but the image will not load.

On my desktop it's:

T, 1, O0, 2, 3, O1, 4, 5, O2, 6, 7, O3, 8, 8*, 9, 10, O4, 11, O5, 11*, 12, 12*, O6, 13, 14, O7, 15, O8, 16, 15*, O9, 17, 18, 18*, 19, O10, 20, 20*, O11.


My desktop did receive all images, but several were double and the duplicate of 15 even came after 16. Wire on my iPhone hasn't stored all images successfully.

Otto responded 12 times in total, I'm assuming that's once to the text and 11 times to one of the images.

This was in a Wi-Fi connection to a weak AP. I then retried the same procedure with Wi-Fi disabled on a better 4G signal. The result is better (I forgot to add a text message, so T/O0 is missing).

Phone: 1, 2, 3, 4, 5, 6, O1, 7, 8, O2, 9, 10, 11, O3, 12, 13, 14, 15, O4, 16, 17, 18, O5, 19, 20, O6
Desktop: 1, 2, 3, 4, O1, 5, 6, 7, 8, O2, 9, 10, 11, O3, 12, 13, 14, O4, 15, 16, 17, O5, 18, 19, 20, O6

So in this situation, both the desktop and the phone have the same messages without duplicates. Otto only responded 6 times in total.

And then I retried adding a text message, and I got similar results on 4G as on W-Fi: some duplicate images, some image in reversed error on desktop.

To make sure it wasn't really about the text message, I tried it again using 4G without a text message, and the errors also occurred there: incorrect order, duplicates, and several images missing in Wire on my iPhone.

**TL/DR**
Unfornately still errors: duplicates, incorrect order, and sometimes sent images show up on my phone with just a loading icon but never load. Additionally, the system sharing window sometimes doesn't close by itself / doesn't finish.

**Comment by MadEgg** [comment]
With the latest version (2.35 (2825)) there actually seems to be a regression. I can barely send images at all through the System Share Extension.

When selecting and sending images from the Photos app, the Wire sending screen often does not close. The text message is sent. The images also appear in my conversation, but are stuck in the 'Sending' state. Shutting down Wire and restarting it sometimes helps, sometimes it doesn't help.

I've experienced this several times now since the last update, and some of my contacts experience the same issue.

**Comment by anatoli26** [comment]
I'm experiencing issues with photo sharing too... can't share more than 5-6 photos/videos at a time (Wire just doesn't appear in the share options), Wire sometimes hangs during the sending process, most of the times I'm getting duplicates, sometimes sharing a photo doesn't show anything in Wire, then after restarting it all the files that I tried to upload without seeing anything in Wire, suddenly appear there uploading...

And of course the problem of showing to the peers the incomplete uploads as something that is already downloading to their devices and never finishes is adding additional confusion and frustration.

**Comment by EgbertW** [comment]
It still isn't working as it should. Over the past weeks when I selected multiple images and sent them to a contact, it didn't send them at all but rather locked them in the 'Sending' state. Shutting down Wire and restarting it caused it to send the first image, and then crash. Restarting Wire caused it to send the next image and crash again, and so on.

**Comment by johnxnguyen** [comment]
Hey @EgbertW! I just want to let you know that we have addressed this issue and the fix will be in an upcoming release 😄

**Comment by EgbertW** [comment]
@johnxnguyen Great to hear, I'll keep checking for new updates and test it as soon as I can.

One of my contacts uses Wire on Android and experiences similar problems - are these things related and is an update for the Android version also upcoming?

**Comment by johnxnguyen** [comment]
@EgbertW FYI I believe it will be released in version 2.43. In regards to the Android client, I will have to get back to you on that one. But it is quite possible that the issues are related. 

On the iOS side, the problem stems from the way the operating system handles app extensions in regard to memory consumption. Basically, if the extension exceeds the strict limits enforced by the OS, then it will be simply terminated. This is why the issue is most noticeable when sending multiple photos, especially if they are at full resolution. I'm not familiar with how the Android OS manages app extensions but I would assume it is quite similar. I will speak to one of the Android developers to get back to you as soon as I can.

**Comment by johnxnguyen** [comment]
@EgbertW so the Android team haven't had this issue reported so if your contact continues to experience problems with sharing then it would be very helpful for them to file an issue on the android GitHub repositories or with Wire customer support.
